File: ass1/sorts_benchmark.go
package main

import (
	"encoding/csv"
	"fmt"
	"math/rand"
	"os"
	"strconv"
	"time"
)

func main() {
	inputSizes := []int{100, 500, 1000, 2000, 5000, 10000}

	numRuns := 5

	file, err := os.Create("benchmark_results.csv")
	if err != nil {
		fmt.Printf("Error creating CSV file: %v\n", err)
		return
	}
	defer file.Close()

	writer := csv.NewWriter(file)
	defer writer.Flush()

	header := []string{"Algorithm", "InputSize", "TimeMs"}
	if err := writer.Write(header); err != nil {
		fmt.Printf("Error writing header: %v\n", err)
		return
	}

	fmt.Println("Running benchmarks...")
	fmt.Println("===================")

	// Benchmark each sorting function
	for _, sortFunc := range sortFunctions {
		fmt.Printf("\nTesting %s:\n", sortFunc.name)

		for _, size := range inputSizes {
			var totalTime time.Duration

			// Run multiple times and average
			for range numRuns {
				arr := generateRandomArray(size)

				start := time.Now()
				err := sortFunc.fn(arr)
				elapsed := time.Since(start)

				if err != nil {
					fmt.Printf("Error in %s with size %d: %v\n", sortFunc.name, size, err)
					continue
				}

				totalTime += elapsed
			}

			// Calculate average time in milliseconds
			avgTime := totalTime / time.Duration(numRuns)
			avgTimeMs := float64(avgTime.Microseconds()) / 1000.0

			record := []string{
				sortFunc.name,
				strconv.Itoa(size),
				fmt.Sprintf("%.4f", avgTimeMs),
			}

			if err := writer.Write(record); err != nil {
				fmt.Printf("Error writing record: %v\n", err)
				continue
			}

			fmt.Printf("  Size %d: %.4f ms (avg of %d runs)\n", size, avgTimeMs, numRuns)
		}
	}

	fmt.Println("\n===================")
	fmt.Println("Benchmark results saved to benchmark_results.csv")
}

func generateRandomArray(size int) []int {
	arr := make([]int, size)
	for i := range size {
		arr[i] = rand.Intn(10000)
	}
	return arr
}

File: ass1/sorts_benchmark.py
import csv
import random
import time
from sorts import insertion_sort, merge_sort, heap_sort, quick_sort


def benchmark_sorts():
    sort_functions = [
        ("InsertionSort", insertion_sort),
        ("MergeSort", merge_sort),
        ("HeapSort", heap_sort),
        ("QuickSort", quick_sort),
    ]

    input_sizes = [100, 500, 1000, 2000, 5000, 10000]

    num_runs = 5

    with open("python_benchmark_results.csv", "w", newline="") as file:
        writer = csv.writer(file)

        writer.writerow(["Algorithm", "InputSize", "TimeMs"])

        print("Running Python benchmarks...")
        print("===================")

        # Benchmark each sorting function
        for name, sort_func in sort_functions:
            print(f"\nTesting {name}:")

            for size in input_sizes:
                total_time = 0.0

                # Run multiple times and average
                for _ in range(num_runs):
                    arr = [random.randint(0, 9999) for _ in range(size)]

                    start = time.perf_counter()
                    sort_func(arr)
                    elapsed = time.perf_counter() - start

                    total_time += elapsed

                # Calculate average time in milliseconds
                avg_time_ms = (total_time / num_runs) * 1000

                writer.writerow([name, size, f"{avg_time_ms:.4f}"])

                print(f"  Size {size}: {avg_time_ms:.4f} ms (avg of {num_runs} runs)")

        print("\n===================")
        print("Benchmark results saved to python_benchmark_results.csv")


if __name__ == "__main__":
    benchmark_sorts()
